owncloud_service/source_watcher.py

```diff
@@ -0,0 +1,298 @@
+import argparse
+import os
+import time
+from datetime import datetime, timedelta, timezone
+
+import requests
+from requests.auth import HTTPBasicAuth
+from slack_bot import setup_logger
+
+# ----- OwnCloud Configuration -----
+BASE_URL = "https://grandma-owncloud.lal.in2p3.fr/remote.php/dav/files"
+OWNCLOUD_USERNAME = "OWNCLOUD_USERNAME"  # Replace with your ownCloud username
+OWNCLOUD_TOKEN = "OWNCLOUD_TOKEN"  # Replace with your ownCloud token
+OWNCLOUD_ID = (
+    "OWNCLOUD_ID"  # Replace with your ownCloud ID (from settings, WebDAV section)
+)
+# -----------------------------------
+
+# ----- Configuration -----
+INSTANCE_URL = "https://skyportal-icare.ijclab.in2p3.fr"  # Instance URL
+API_TOKEN = "API_TOKEN"  # Replace with your API token
+SAVE_PATH = (
+    "Candidates/Skyportal"  # Base directory where source folders will be created
+)
+SOURCE_TAG = ""  # Optional tag to filter sources
+POLL_INTERVAL = 60  # Polling interval in seconds
+SKYPORTAL_GROUP_IDS_FILTER = [
+    3
+]  # save only sources saved to one of these skyportal groups IDs (GRANDNMA group ID is 3)
+USE_BASE_TELESCOPE_LIST = True  # Use predefined telescope list or fetch from SkyPortal
+# ----------------------------------
+
+TELESCOPE_LIST = [
+    "TAROT-TCA",
+    "TAROT-TRE",
+    "TAROT-TCH",
+    "Les-Makes-T60",
+    "UBAI-NT-60",
+    "UBAI-ST-60",
+    "FRAM-CTA-N",
+    "FRAM-Auger",
+    "OHP-IRIS",
+    "AbAO-T150",
+    "VIRT",
+    "TRT-SBO",
+    "TRT-GAO",
+    "TRT-SRO",
+    "TRT-CTO",
+    "TNT",
+    "ShAO-T60",
+    "AbAO-T70",
+    "GMG-2.4",
+    "Xinglong-2.16m",
+    "OST-CDK",
+    "HAO",
+    "KAO",
+    "OPD-60cm",
+]
+
+header = {}
+seen_sources = set()  # Track seen source ids to avoid duplication
+
+
+def get_new_sources(start_time, last_iteration):
+    """Fetch new sources from SkyPortal with an optional tag filter."""
```

Comment:
It would be good to add type annotation + docstring arguments declaration on all functions.